File: assemblyscript-algorithms/src/algorithms/search/interpolation-search.ts
function _interpolationSearch_u64_asm(
  array: StaticArray<u64>,
  startOffset: i32,
  endOffset: i32,
  target: u64
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        <i32>(array[endOffset] - array[startOffset])) *
        <i32>(target - array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_u64_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_u64_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_u64_asm(
  array: StaticArray<u64>,
  target: u64
): i32 {
  return _interpolationSearch_u64_asm(array, 0, array.length - 1, target);
}
function _interpolationSearch_i64_asm(
  array: StaticArray<i64>,
  startOffset: i32,
  endOffset: i32,
  target: i64
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        <i32>(array[endOffset] - array[startOffset])) *
        <i32>(target - array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_i64_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_i64_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_i64_asm(
  array: StaticArray<i64>,
  target: i64
): i32 {
  return _interpolationSearch_i64_asm(array, 0, array.length - 1, target);
}
function _interpolationSearch_f64_asm(
  array: StaticArray<f64>,
  startOffset: i32,
  endOffset: i32,
  target: f64
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        <i32>floor(array[endOffset] - array[startOffset])) *
        <i32>floor(target - array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_f64_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_f64_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_f64_asm(
  array: StaticArray<f64>,
  target: f64
): i32 {
  return _interpolationSearch_f64_asm(array, 0, array.length - 1, target);
}
function _interpolationSearch_f32_asm(
  array: StaticArray<f32>,
  startOffset: i32,
  endOffset: i32,
  target: f32
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        <i32>floor(array[endOffset] - array[startOffset])) *
        <i32>floor(target - array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_f32_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_f32_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_f32_asm(
  array: StaticArray<f32>,
  target: f32
): i32 {
  return _interpolationSearch_f32_asm(array, 0, array.length - 1, target);
}
function _interpolationSearch_u32_asm(
  array: StaticArray<u32>,
  startOffset: i32,
  endOffset: i32,
  target: u32
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        <i32>(array[endOffset] - array[startOffset])) *
        <i32>(target - array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_u32_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_u32_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_u32_asm(
  array: StaticArray<u32>,
  target: u32
): i32 {
  return _interpolationSearch_u32_asm(array, 0, array.length - 1, target);
}
function _interpolationSearch_i32_asm(
  array: StaticArray<i32>,
  startOffset: i32,
  endOffset: i32,
  target: i32
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
        (target - array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_i32_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_i32_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_i32_asm(
  array: StaticArray<i32>,
  target: i32
): i32 {
  return _interpolationSearch_i32_asm(array, 0, array.length - 1, target);
}

function _interpolationSearch_u16_asm(
  array: StaticArray<u16>,
  startOffset: i32,
  endOffset: i32,
  target: u16
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        (<i32>array[endOffset] - <i32>array[startOffset])) *
        (<i32>target - <i32>array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_u16_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_u16_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_u16_asm(
  array: StaticArray<u16>,
  target: u16
): i32 {
  return _interpolationSearch_u16_asm(array, 0, array.length - 1, target);
}
function _interpolationSearch_i16_asm(
  array: StaticArray<i16>,
  startOffset: i32,
  endOffset: i32,
  target: i16
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        (<i32>array[endOffset] - <i32>array[startOffset])) *
        (<i32>target - <i32>array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_i16_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_i16_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_i16_asm(
  array: StaticArray<i16>,
  target: i16
): i32 {
  return _interpolationSearch_i16_asm(array, 0, array.length - 1, target);
}
function _interpolationSearch_u8_asm(
  array: StaticArray<u8>,
  startOffset: i32,
  endOffset: i32,
  target: u8
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        (<i32>array[endOffset] - <i32>array[startOffset])) *
        (<i32>target - <i32>array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_u8_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_u8_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_u8_asm(
  array: StaticArray<u8>,
  target: u8
): i32 {
  return _interpolationSearch_u8_asm(array, 0, array.length - 1, target);
}
function _interpolationSearch_i8_asm(
  array: StaticArray<i8>,
  startOffset: i32,
  endOffset: i32,
  target: i8
): i32 {
  let pos = -1;

  if (
    startOffset <= endOffset &&
    target >= array[startOffset] &&
    target <= array[endOffset]
  ) {
    pos =
      startOffset +
      ((endOffset - startOffset) /
        (<i32>array[endOffset] - <i32>array[startOffset])) *
        (<i32>target - <i32>array[startOffset]);

    if (array[pos] == target) {
      return pos;
    }

    if (array[pos] < target) {
      return _interpolationSearch_i8_asm(array, pos + 1, endOffset, target);
    }

    if (array[pos] > target) {
      return _interpolationSearch_i8_asm(array, startOffset, pos - 1, target);
    }
  }
  return -1;
}
export function interpolationSearch_i8_asm(
  array: StaticArray<i8>,
  target: i8
): i32 {
  return _interpolationSearch_i8_asm(array, 0, array.length - 1, target);
}

File: c-algorithms/src/search/interpolation-search/interpolation-search.cpp
#include <math.h>
#include "search/interpolation-search/interpolation-search.h";
#include "shared/shared.h"

int _interpolationSearch_u64_c(u64* array, int startOffset, int endOffset, u64 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset]);

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_u64_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_u64_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_u64_c(struct Array1d_u64* array, u64 target) {
    return _interpolationSearch_u64_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_i64_c(i64* array, int startOffset, int endOffset, i64 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset]);

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_i64_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_i64_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_i64_c(struct Array1d_i64* array, i64 target) {
    return _interpolationSearch_i64_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_f64_c(f64* array, int startOffset, int endOffset, f64 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
            floor(
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset])
            );

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_f64_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_f64_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_f64_c(struct Array1d_f64* array, f64 target) {
    return _interpolationSearch_f64_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_f32_c(f32* array, int startOffset, int endOffset, f32 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
            floor(
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset])
            );

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_f32_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_f32_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_f32_c(struct Array1d_f32* array, f32 target) {
    return _interpolationSearch_f32_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_u32_c(u32* array, int startOffset, int endOffset, u32 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset]);

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_u32_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_u32_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_u32_c(struct Array1d_u32* array, u32 target) {
    return _interpolationSearch_u32_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_i32_c(i32* array, int startOffset, int endOffset, i32 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset]);

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_i32_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_i32_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_i32_c(struct Array1d_i32* array, i32 target) {
    return _interpolationSearch_i32_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_u16_c(u16* array, int startOffset, int endOffset, u16 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset]);

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_u16_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_u16_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_u16_c(struct Array1d_u16* array, u16 target) {
    return _interpolationSearch_u16_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_i16_c(i16* array, int startOffset, int endOffset, i16 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset]);

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_i16_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_i16_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_i16_c(struct Array1d_i16* array, i16 target) {
    return _interpolationSearch_i16_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_u8_c(u8* array, int startOffset, int endOffset, u8 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset]);

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_u8_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_u8_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_u8_c(struct Array1d_u8* array, u8 target) {
    return _interpolationSearch_u8_c(array->items, 0, array->length - 1, target);
}
int _interpolationSearch_i8_c(i8* array, int startOffset, int endOffset, i8 target) {
    int pos = -1;
    if (
        startOffset <= endOffset &&
        target >= array[startOffset] &&
        target <= array[endOffset]
    ) {
        pos =
            startOffset +
                ((endOffset - startOffset) / (array[endOffset] - array[startOffset])) *
                (target - array[startOffset]);

        if (array[pos] == target) {
            return pos;
        }

        if (array[pos] < target) {
            return _interpolationSearch_i8_c(array, pos + 1, endOffset, target);
        }

        if (array[pos] > target) {
            return _interpolationSearch_i8_c(array, startOffset, pos - 1, target);
        }
    }

    return -1;
}
int interpolationSearch_i8_c(struct Array1d_i8* array, i8 target) {
    return _interpolationSearch_i8_c(array->items, 0, array->length - 1, target);
}
